certs/certs.go

```diff
@@ -0,0 +1,156 @@
+// Package certs handles persisting and loading certs
+package certs
+
+import (
+	"bufio"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"github.com/levenlabs/locksmith/config"
+	"log"
+	"os"
+	"time"
+)
+
+type saveReq struct {
+	Hostname    string
+	RemoteAddr  string
+	Certificate string
+	ReplyCh     chan error
+}
+
+type Certificate struct {
+	Hostname    string `json:"hostname"`
+	RemoteAddr  string `json:"remoteAddr"`
+	Generated   int64  `json:"generated_at"`
+	Certificate string `json:"certificate"`
+}
+
+type listReq struct {
+	ReplyCh chan []Certificate
+}
+
+var (
+	listCh  chan listReq
+	saveCh  chan saveReq
+	timeFmt = "2006-01-02T15:04:05"
+	fileFmt = "%s | %s | %s | %s"
+)
+
+func init() {
+	if config.CertsFile != "" {
+		// make sure we can write to the file
+		f, err := os.OpenFile(config.CertsFile, os.O_CREATE, 0600)
+		if err != nil {
+			log.Fatalf("Failed to create certs file: %s", err)
+		}
+		f.Close()
+	}
+
+	listCh = make(chan listReq)
+	saveCh = make(chan saveReq)
+	go func() {
+		var l listReq
+		var s saveReq
+		for {
+			select {
+			case l = <-listCh:
+				l.ReplyCh <- listCerts()
+			case s = <-saveCh:
+				s.ReplyCh <- saveCert(s.Hostname, s.RemoteAddr, s.Certificate)
+			}
+		}
+	}()
+}
+
+func SaveCert(hostname, remoteAddr, cert string) error {
+	r := make(chan error)
+	saveCh <- saveReq{
+		Hostname:    hostname,
+		RemoteAddr:  remoteAddr,
+		Certificate: cert,
+		ReplyCh:     r,
+	}
+	return <-r
+}
+
+func saveCert(hostname, remoteAddr, cert string) error {
+	if config.CertsFile == "" {
+		return nil
+	}
+
+	b, _ := pem.Decode([]byte(cert))
+	if b == nil || len(b.Bytes) == 0 {
+		return errors.New("Error reading PEM certificate")
+	}
+
+	b64cert := base64.StdEncoding.EncodeToString(b.Bytes)
+	t := time.Now()
```

Comment:
If you're not ever using `t` you can chain this like `tf := time.Now().Format(timeFmt)`